File: lib/providers/browser_provider.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../utils/common_utils.dart';

class BrowserState {
  static const String defaultYouTubeUrl = 'https://m.youtube.com';

  final String currentUrl;
  final bool isYoutube;
  final bool isLoading;
  final double progress;
  final bool isInitialLoad;
  final bool hasShownError;

  const BrowserState({
    this.currentUrl = defaultYouTubeUrl,
    this.isYoutube = true,
    this.isLoading = false,
    this.progress = 0.0,
    this.isInitialLoad = true,
    this.hasShownError = false,
  });

  BrowserState copyWith({
    String? currentUrl,
    bool? isYoutube,
    bool? isLoading,
    double? progress,
    bool? isInitialLoad,
    bool? hasShownError,
  }) {
    return BrowserState(
      currentUrl: currentUrl ?? this.currentUrl,
      isYoutube: isYoutube ?? this.isYoutube,
      isLoading: isLoading ?? this.isLoading,
      progress: progress ?? this.progress,
      isInitialLoad: isInitialLoad ?? this.isInitialLoad,
      hasShownError: hasShownError ?? this.hasShownError,
    );
  }
}

class BrowserNotifier extends Notifier<BrowserState> {
  @override
  BrowserState build() {
    return const BrowserState();
  }

  bool _isYouTubeUrl(String url) {
    return YouTubeUtils.isYouTubeUrl(url);
  }

  void setProgress(double progress) {
    state = state.copyWith(progress: progress);
  }

  void setUrl(String url) {
    state = state.copyWith(
      currentUrl: url,
      isYoutube: _isYouTubeUrl(url),
    );
  }

  void setLoading(bool isLoading) {
    state = state.copyWith(isLoading: isLoading);
  }

  void setInitialLoad(bool isInitialLoad) {
    state = state.copyWith(isInitialLoad: isInitialLoad);
  }

  void setHasShownError(bool hasShownError) {
    state = state.copyWith(hasShownError: hasShownError);
  }
}

final browserProvider =
    NotifierProvider.autoDispose<BrowserNotifier, BrowserState>(
  BrowserNotifier.new,
);

File: android/app/src/main/kotlin/com/example/ytdown/providers/BrowserProvider.kt
package com.example.ytdown.providers

import com.example.ytdown.utils.YouTubeUtils
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

data class BrowserState(
    val currentUrl: String = BrowserState.defaultYouTubeUrl,
    val isYoutube: Boolean = true,
    val isLoading: Boolean = false,
    val progress: Double = 0.0,
    val isInitialLoad: Boolean = true,
    val hasShownError: Boolean = false
) {
    companion object {
        const val defaultYouTubeUrl = "https://m.youtube.com"
    }

    fun copyWith(
        currentUrl: String? = null,
        isYoutube: Boolean? = null,
        isLoading: Boolean? = null,
        progress: Double? = null,
        isInitialLoad: Boolean? = null,
        hasShownError: Boolean? = null
    ): BrowserState {
        return BrowserState(
            currentUrl = currentUrl ?: this.currentUrl,
            isYoutube = isYoutube ?: this.isYoutube,
            isLoading = isLoading ?: this.isLoading,
            progress = progress ?: this.progress,
            isInitialLoad = isInitialLoad ?: this.isInitialLoad,
            hasShownError = hasShownError ?: this.hasShownError
        )
    }
}

class BrowserProvider {
    private val _state = MutableStateFlow(BrowserState())
    val state: StateFlow<BrowserState> = _state.asStateFlow()

    fun setUrl(url: String) {
        // Apenas atualiza a URL e o estado Youtube, sem resetar isInitialLoad ou hasShownError
        _state.value = _state.value.copyWith(
            currentUrl = url,
            isYoutube = isYouTubeUrl(url),
            progress = 0.0
        )
    }

    fun setLoading(isLoading: Boolean) {
        _state.value = _state.value.copyWith(isLoading = isLoading)
    }

    fun setProgress(progress: Double) {
        _state.value = _state.value.copyWith(progress = progress)
    }

    fun setInitialLoad(isInitialLoad: Boolean) {
        _state.value = _state.value.copyWith(isInitialLoad = isInitialLoad)
    }

    fun setHasShownError(hasShownError: Boolean) {
        _state.value = _state.value.copyWith(hasShownError = hasShownError)
    }

    fun reset() {
        _state.value = BrowserState()
    }

    private fun isYouTubeUrl(url: String): Boolean {
        return YouTubeUtils.isYouTubeUrl(url)
    }
}

val browserProvider = BrowserProvider()
